Src/ILGPU.Algorithms/ConcurrentStreamProcessor.cs

```diff
@@ -0,0 +1,161 @@
+        #region Instance
+
+        private readonly AcceleratorStream[] streams;
+
+        /// <summary>
+        /// Constructs a new concurrent stream processor.
+        /// </summary>
+        /// <param name="accelerator">The parent accelerator.</param>
+        /// <param name="maxNumConcurrentStreams">
+        /// The maximum number of concurrent streams to use (if any).
+        /// </param>
+        /// <param name="streamProvider">
+        /// A custom stream provider function to construct specialized streams.
+        /// </param>
+        public ConcurrentStreamProcessor(
+            Accelerator accelerator,
+            int maxNumConcurrentStreams = 0,
+            Func<Accelerator, AcceleratorStream> streamProvider = null)
+        {
+            maxNumConcurrentStreams = Math.Max(
+                Math.Max(maxNumConcurrentStreams, 1),
+                Environment.ProcessorCount);
+
+            streams = new AcceleratorStream[maxNumConcurrentStreams];
+            if (maxNumConcurrentStreams > 1)
+            {
+                streamProvider ??= accel => accel.CreateStream();
+                for (int i = 0; i < streams.Length; ++i)
+                    streams[i] = streamProvider(accelerator);
+            }
+        }
+
+        #endregion
+
+        #region Properties
+
+        /// <summary>
+        /// Returns the maximum number of concurrent streams supported by this processor.
+        /// </summary>
+        public int MaxNumConcurrentStreams => streams.Length;
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Processes the given action concurrently by using the underlying accelerator
+        /// streams concurrently to submit different jobs in parallel.
+        /// </summary>
+        /// <remarks>
+        /// Note that this method assumes that all previous jobs have been synchronized
+        /// with the current processing thread.
+        /// </remarks>
+        /// <param name="numActions">The number of actions to submit.</param>
+        /// <param name="action">
+        /// The action to invoke on each stream to submit work.
+        /// </param>
+        public void ProcessConcurrently(
+            int numActions,
+            Action<AcceleratorStream, int> action) =>
+            ProcessConcurrently(null, numActions, action);
+
+        /// <summary>
+        /// Processes the given action concurrently by using the underlying accelerator
+        /// streams concurrently to submit different jobs in parallel.
+        /// </summary>
+        /// <param name="stream">The current accelerator stream.</param>
+        /// <param name="numActions">The number of actions to submit.</param>
+        /// <param name="action">
+        /// The action to invoke on each stream to submit work.
+        /// </param>
+        public void ProcessConcurrently(
+            AcceleratorStream stream,
+            int numActions,
+            Action<AcceleratorStream, int> action)
```

Comment:
Apologies, but I feel like this design is probably not what you want. It looks to me like you have a fair amount of locks here and in the case where you have a few long running actions this will bottleneck and stall while it waits for an action to finish.  Basically, what I think you want to do is:
1. Not tie this to a particular individual stream.
2. Dispatch this to a waiting stream, if available.
3. If no stream available wait until one signals ready.

Reply:
I'm a .NET beginner, but it feels to me like you would want to make this similar to the existing [system.Threading.ThreadPool].
https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadpool?view=net-7.0
https://referencesource.microsoft.com/#mscorlib/system/threading/threadpool.cs

Reply:
Thanks for your comments. It will indeed distribute the workload by attaching them to the the different accelerator streams available. This way, we can launch multiple kernels in parallel on the device. One design decision could be to also use the incoming stream for further processing. However, this could lead to invalid results as all previous operations on the device need to be finished before using multiple async streams. The downside of this approach is clearly the fact that we have to wait for the input accelerator stream to finish. There is no other way to launch multiple kernels concurrently on the device.